tunnel-server/src/app.ts

```diff
@@ -31,4 +40,37 @@ export const app = ({ isProxyRequest, proxyHandlers, logger }: {
     logger,
   })
     .register(fastifyRequestContext)
+    .get<{Querystring: {env: string, returnPath?: string}}>('/login', {
+      schema: {
+        querystring:{
+          properties: {
+            env: { type: 'string' },
+            returnPath: { type: 'string' }
+          },
+          required: ['env']
+        }
+      }
+    }, async function (req, res) {
+      const {env:envId, returnPath = '/'} = req.query
+      if (!returnPath.startsWith('/')){
+        res.statusCode = 400
+        return {error: 'returnPath must be a relative path' }
+      }
+      const env = await envStore.get(envId)
+      if (!env){
+        res.statusCode = 404
+        return {error: 'unknown envId'}
+      }
+      const session = sessionManager(req.raw, res.raw, env.publicKeyThumbprint)
+      if (!session.user){
+        const auth = authenticator([JwtAuthenticator(env)])
+        const result = await auth(req.raw)
+        if (!result.isAuthenticated){
+          return unauthorized(res.raw)
+        }
+        session.set(result.claims)
+        session.save()
+      }
+      return res.redirect(new URL(returnPath, replaceHostname(baseUrl, `${envId}.${baseUrl.hostname}`)).toString())
```

Comment:
missing `clientId`?

Reply:
I think the envId contains the clientId (since it's basically the host we're comparing).
We need to change the name of envStore, it's not really env it's a hostname->socket/tunnel